lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -168,36 +182,158 @@ data BalanceInsufficientError = BalanceInsufficientError
+
+        makeChangeRepeatedly preSelection leftover <&> \case
+            Nothing -> -- Running out of ada-only UTxO to pay for fee / deposits
+                Left UnableToConstructChange
+            Just result ->
+                Right result
   where
+    SelectionCriteria{outputsToCover, utxoAvailable, extraCoinSource} = criteria
+
+    -- Outputs require a minimum ada value (to avoid filling the blockchain with
+    -- worthless UTxOs, any UTxO is required to hold some Ada which gives it an
+    -- implicit value).
+    --
+    -- However, users do not typically specify a minimum ada value themselves.
+    -- One would rather send '10 Apple' and not '10 Apple & 1.2 Ada'. Therefore,
+    -- unless a coin value is explicitely specified, we do assign a coin value
+    -- manually for each non-ada output. That value is the minimum value
+    -- possible to make a particular output valid.
+    outputsConsidered :: NonEmpty TxOut
+    outputsConsidered = flip fmap outputsToCover $ \out ->
+        out { Tx.tokens = augmentBundle (Tx.tokens out) }
+      where
+        augmentBundle bundle =
+            if TokenBundle.getCoin bundle == Coin 0 then
+                bundle { coin = minCoinValueFor (view #tokens bundle) }
+            else
+                bundle
+
+    mkInputsSelected :: UTxOIndex -> NonEmpty (TxIn, TxOut)
+    mkInputsSelected =
+        fromMaybe invariantSelectAnyInputs . NE.nonEmpty . UTxOIndex.toList
+
     balanceAvailable :: TokenBundle
     balanceAvailable = view #balance utxoAvailable
 
     balanceRequired :: TokenBundle
-    balanceRequired = F.foldMap (view #tokens) outputsToCover
-
-    mkResult :: SelectionState -> SelectionResult
-    mkResult SelectionState {selected, leftover} =
-        case NE.nonEmpty (UTxOIndex.toList selected) of
-            Nothing ->
-                unableToSelectAnyInputsError
-            Just inputsSelected ->
-                SelectionResult
+    balanceRequired = F.foldMap (view #tokens) outputsConsidered
+
+    -- There's a chicken-and-egg situation when it comes to calculating
+    -- transaction fees. On the one hand, we need to know the shape of the final
+    -- transaction to calculate its cost. But in order to construct the
+    -- transaction, we need to know what its cost is.
+    --
+    -- So, in order to not duplicate the logic from 'makeChange', we first
+    -- calculate a pre-selection considering the case where we have no fees to
+    -- pay, and no minimum value. This is guaranteed to succeed and will yield
+    -- the final shape of the change outputs (modulo amounts).
+    --
+    -- From there, we can successively re-iterate and construct a final
+    -- unbalanced selection that is leaving enough money out for fees. For this
+    -- method to work (efficiently), the function estimate the cost of a
+    -- selection MUST NOT depends on change token quantity. That is, adding more
+    -- change to an output should not make it more expensive.
```

Comment:
> Does it mean

It means precisely that indeed! This may be confusing unless you have a deeper look / understanding about how fees are actually calculated. This has to do with the way objects are encoded to binary with CBOR. CBOR uses a variable-length encoding for numbers. That is, small numbers are encoded with 1 byte, bigger number with 3, or 5 or 9. 

Our fee calculation function now often consider the worse case to avoid the hassle coming from being _too precise_. Dealing with this has been a nightmare for a long time, everytime we would increase the quantity of a change output, fees would also increase, and as a result, we would need to reduce change outputs, and fees would decrease... and so forth ad infinitum.

Reply:
Understood!

In that case, I think a simple tweak to the comment would clear up any confusion:

```patch
- adding more change to an output should not make it more expensive.
+ increasing the token quantities of a change output should not make it more expensive.
```